clang/test/Sema/attr-nonblocking-constraints.cpp

```diff
@@ -0,0 +1,194 @@
+			return {};
+		}
+	};
+
+	struct Stringy2 {
+		friend Stringy2 operator+(const Stringy2& x, const Stringy2& y)
+		{
+			// Do something inferably unsafe
+			throw 42; // expected-note {{function cannot be inferred 'nonblocking' because it throws or catches exceptions}}
+		}
+	};
+
+void nl9() [[clang::nonblocking]]
+{
+	Adder<int>::add_explicit(1, 2);
+	Adder<int>::add_implicit(1, 2);
+
+	Adder<Stringy>::add_explicit({}, {}); // expected-note {{in template expansion here}}
+	Adder<Stringy2>::add_implicit({}, {}); // expected-warning {{'nonblocking' function must not call non-'nonblocking' function}} \
+		expected-note {{in template expansion here}}
+}
+
+void nl10(
+	void (*fp1)(), // expected-note {{function pointer cannot be inferred 'nonblocking'}}
+	void (*fp2)() [[clang::nonblocking]]
+	) [[clang::nonblocking]]
+{
+	fp1(); // expected-warning {{'nonblocking' function must not call non-'nonblocking' function}}
+	fp2();
+}
+
+// Interactions with nonblocking(false)
+void nl11_no_inference_1() [[clang::nonblocking(false)]] // expected-note {{function does not permit inference of 'nonblocking'}}
+{
+}
+void nl11_no_inference_2() [[clang::nonblocking(false)]]; // expected-note {{function does not permit inference of 'nonblocking'}}
+
+template <bool V>
+struct ComputedNB {
+	void method() [[clang::nonblocking(V)]]; // expected-note {{function does not permit inference of 'nonblocking'}}
+};
+
+void nl11() [[clang::nonblocking]]
+{
+	nl11_no_inference_1(); // expected-warning {{'nonblocking' function must not call non-'nonblocking' function}}
+	nl11_no_inference_2(); // expected-warning {{'nonblocking' function must not call non-'nonblocking' function}}
+
+	ComputedNB<true> CNB_true;
+	CNB_true.method();
+	
+	ComputedNB<false> CNB_false;
+	CNB_false.method(); // expected-warning {{'nonblocking' function must not call non-'nonblocking' function}}
+}
+
+// Verify that when attached to a redeclaration, the attribute successfully attaches.
+void nl12() {
+	static int x; // expected-warning {{'nonblocking' function must not have static locals}}
+}
+void nl12() [[clang::nonblocking]];
+void nl13() [[clang::nonblocking]] { nl12(); }
+
+// C++ member function pointers
+struct PTMFTester {
+	typedef void (PTMFTester::*ConvertFunction)() [[clang::nonblocking]];
+
+	void convert() [[clang::nonblocking]];
+
+	ConvertFunction mConvertFunc;
+};
+
+void PTMFTester::convert() [[clang::nonblocking]]
+{
+	(this->*mConvertFunc)();
+}
+
+// Block variables
+void nl17(void (^blk)() [[clang::nonblocking]]) [[clang::nonblocking]] {
+	blk();
+}
+
+// References to blocks
+void nl18(void (^block)() [[clang::nonblocking]]) [[clang::nonblocking]]
+{
+	auto &ref = block;
+	ref();
+}
+
+
+// --- nonblocking implies noexcept ---
+#pragma clang diagnostic warning "-Wperf-constraint-implies-noexcept"
+
+void nl19() [[clang::nonblocking]] // expected-warning {{'nonblocking' function should be declared noexcept}}
+{
+}
```

Comment:
More horrible test cases:
```c++
bool bad();

template <bool>
requires requires { bad(); }
void g() [[clang::nonblocking]] {}

void g() [[clang::nonblocking]] {
    decltype(bad()) a;
    noexcept(bad());
    sizeof(bad());
    [[assume(bad())]]; // never evaluated, but maybe still semantically questionable?
}

void h(bool y = bad()) [[clang::nonblocking]] {
    h(true) // maybe ok? should nonblocking functions be allowed to have blocking default args?
    h(); // warning!
}

// These need '-fms-extensions' (and maybe '-fdeclspec'), so you might
// want to put them in a separate file.
void f1() [[clang::nonblocking]] {
    __try {} __except (1) {}
}

struct S {
    int get_x();
    __declspec(property(get = get_x)) int x;
};

void f2() [[clang::nonblocking]] {
    S a;
    a.x;
}
```

Reply:
> More horrible test cases:
> 
> ```c++
> bool bad();
> 
> template <bool>
> requires requires { bad(); }
> void g() [[clang::nonblocking]] {}
> ```

I'm probably just thick today (and weak on C++20 concepts) but I would have thought that this would fail simply because `bad()` isn't `constexpr`.

Reply:
> I would have thought that this would fail simply because bad() isn't constexpr, but it compiles without diagnostics.

That just checks whether `bad()` is a valid expression. This is mainly about the fact that requires expressions are never evaluated so even if one of them contains a blocking function, that doesn’t make the function containing the expression blocking. This probably already works just fine, but still.

Reply:
> void g() [[clang::nonblocking]] {
>     decltype(bad()) a;
>     noexcept(bad());
>    sizeof(bad());
>     [[assume(bad())]]; // never evaluated, but maybe still semantically questionable?
> }

None of these are visited by the `RecursiveASTVisitor`, for reasons I don't understand. So we don't get the warnings we might expect. But if I use `-c` to generate object code, the object code contains no function call. No harm no foul?

Reply:
Oh! so maybe you're just saying we should have these cases covered by a test but not expect any diagnostics?

Reply:
Yeah, these are mainly just a sanity check to make sure we don’t diagnose unevaluated expressions.

Reply:
Sorry if I didn’t make that too clear...

Reply:
No worries.

Only thing remaining here, then, is that my Apple target doesn't support the MS extensions so I don't have an easy way to work on that last set of checks. Can we make that a follow-on?

Reply:
your clang supports them, you just need to pass an appropriate `--target=` in the `RUN: ` line of the test. The vast majority of clang and llvm's tests can be checked on any host, since we don't tend to write _execution_ tests in this part of the regression suite.

Reply:
Ah, thanks Jon. That works.

Reply:
Yeah, afaik you don’t even need the target (but it’s probably still a good idea to include it); the main thing you need to include in the run line is `-fms-extensions`.